packages/grafana-ui/src/components/uPlot/plugins/TooltipPlugin.tsx

```diff
@@ -142,15 +141,52 @@ export const TooltipPlugin: React.FC<TooltipPluginProps> = ({
     tooltip = <SeriesTable series={series} timestamp={xVal} />;
   }
 
-  if (!tooltip || !coords) {
-    return null;
-  }
-
   return (
     <Portal>
-      <VizTooltipContainer position={{ x: coords.viewport.x, y: coords.viewport.y }} offset={{ x: 10, y: 10 }}>
-        {tooltip}
-      </VizTooltipContainer>
+      {tooltip && coords && (
+        <VizTooltipContainer position={{ x: coords.x, y: coords.y }} offset={{ x: TOOLTIP_OFFSET, y: TOOLTIP_OFFSET }}>
+          {tooltip}
+        </VizTooltipContainer>
+      )}
     </Portal>
   );
 };
+
+function isCursourOutsideCanvas({ left, top }: uPlot.Cursor, canvas: DOMRect) {
+  if (left === undefined || top === undefined) {
+    return false;
+  }
+  return left < 0 || left > canvas.width || top < 0 || top > canvas.height;
+}
+
+/**
+ * Given uPlot cursor position, figure out position of the tooltip withing the canvas bbox
+ * Tooltip is positioned relatively to a viewport
+ * @internal
+ **/
+export function positionTooltip(u: uPlot, bbox: DOMRect) {
+  let x, y;
+  const cL = u.cursor.left || 0;
+  const cT = u.cursor.top || 0;
+
+  if (isCursourOutsideCanvas(u.cursor, bbox)) {
+    const idx = u.posToIdx(cL);
+    // when cursor outside of uPlot's canvas
+    if (cT < 0 || cT > bbox.height) {
+      let pos = findMidPointYPosition(u, idx);
+
+      if (pos) {
+        y = bbox.top + pos;
+        if (cL >= 0 && cL <= bbox.width) {
+          // find x-scale position for a current cursor left position
+          x = bbox.left + u.valToPos(u.data[0][u.posToIdx(cL)], u.series[0].scale!);
+        }
+      }
+    }
+  } else {
+    x = bbox.left + cL;
+    y = bbox.top + cT;
+  }
+
+  return { x, y };
+}
```

Comment:
Unit testing this would be awesome, but not sure how to approach uPlot dependency in this function. Any suggestions @leeoniya ?